177 RST stubs for ReadTheDocs

bbean23 opened the pull request:
## Purpose

_Please briefly describe the purpose of this pull request. If this fixes a github issue please include `e.g. fixes #22` so that the issue gets closed automatically when this is merged._

## Summary of changes

- provided RST stubs for SpotAnalysis classes
- updated docstrings
- added classes to test_DocStringsExist.py

## Implementation notes

Added AutoSummary to Sphinx for generating class name lists at the top of the documentation files.

## Submission checklist
- [x] Target branch is `develop`, not `main`
- [x] Existing tests are updated or new tests were added
- [x] `opencsp/test/test_DocStringsExist.py` are verified to include this change or have been updated accordingly
- [x] .rst file(s) under `doc/` are verified to include this change or have been updated accordingly

## Additional information

none

e10harvey commented on opencsp/common/lib/cv/spot_analysis/SpotAnalysisOperablesStream.py:
```diff
@@ -4,20 +4,43 @@
         self.default_data: SpotAnalysisOperable = None
 
     def set_defaults(self, default_support_images: dict[ImageType, CacheableImage], default_data: SpotAnalysisOperable):
+        """
```
Missing parameter documentation in doc string

e10harvey commented on doc/source/library_reference/library/SpotAnalysis/base_classes.rst:
```diff
@@ -0,0 +1,52 @@
+   :undoc-members:
+   :show-inheritance:
+   :member-order: groupwise
+
```
```suggestion

.. automodule:: opencsp.common.lib.cv.spot_analysis.SpotAnalysisOperablesStream
   :members:
   :undoc-members:
   :show-inheritance:
   :member-order: groupwise

```

e10harvey replied:
I would be inclined to include all non-protected and non-private entities in the documentation

bbean23 replied:
Todo: indicate that this class is not meant for use by users, and most of the time not for use by developers. It is necessary for how the SpotAnalysis class works but only users who really want a deep dive should take a look. Maybe _SpotAnalysisOperablesStream or __SpotAnalysisOperablesStream?

bbean23 replied:
Done!

e10harvey commented on opencsp/common/lib/cv/spot_analysis/image_processor/FalseColorImageProcessor.py:
```diff
@@ -36,6 +36,12 @@ def __init__(self, map_type='human', opencv_map=cv2.COLORMAP_JET):
         self.opencv_map = opencv_map
 
     def apply_mapping_jet_custom(self, operable: SpotAnalysisOperable) -> SpotAnalysisOperable:
+        """Updates the primary image with a false color map ('human' or
```
Missing parameter docs for operable

e10harvey commented on doc/source/library_reference/library/SpotAnalysis/image_processor.rst:
```diff
@@ -0,0 +1,141 @@
+
+   AbstractAggregateImageProcessor
+   AbstractSpotAnalysisImageProcessor
+   AbstractVisualizationImageProcessor
```
```suggestion
   AbstractSpotAnalysisImageProcessorLeger
   AbstractVisualizationImageProcessor
```

e10harvey commented on opencsp/test/test_DocStringsExist.py:
```diff
@@ -1,28 +1,70 @@
+                    undocumented_methods.append(method)
+
+            self.assertEqual(
+                len(undocumented_methods), 0, "Found undocumented methods:\n\t" + "\n\t".join(undocumented_methods)
```
thank you

e10harvey reviewed (changes requested):
Thank you for these changes @bbean23. I left some minor comments below. Please reorganize these rst files to match the directory structore of the code base, like this:
`doc/source/library_reference/library/SpotAnalysis -> doc/source/library_reference/common/lib/cv/spot_analysis`. I'd be happy to open a PR into your fork with these changes.

bbean23 reviewed (commented):
I believe I've addressed all your comments. Please note that by "addressed" I sometimes mean "argued against", so please take a look at my comments to see if you agree or not.

e10harvey commented on doc/source/library_reference/library/SpotAnalysis/base_classes.rst:
```diff
@@ -0,0 +1,52 @@
+SpotAnalysis
+============
```
I don't have a strong preference but the current documentation uses the second style

e10harvey commented on doc/source/library_reference/library/SpotAnalysis/index.rst:
```diff
@@ -0,0 +1,13 @@
+   :maxdepth: 1
+
+   base_classes.rst
+   image_processor.rst
```
Either way is fine with me but I want to make sure both the code and rst is maintainable and organized. Maybe the compromise is appending `config_` to base_classes.rst and image_processor.rst.

bbean23 replied:
I'm not sure what you mean by this. You mean changing the file name to `base_classes_config.rst` and `image_processor_config.rst`?

bbean23 replied:
To be addressed in follow-on issue

e10harvey reviewed (commented):
Thank you @bbean23! I think this can be merged after the headings are made consistent with the rest of the library reference.

e10harvey commented on opencsp/common/lib/cv/SpotAnalysis.py:
```diff
@@ -34,85 +35,117 @@ class SpotAnalysis(Iterator[tuple[SpotAnalysisOperable]]):
+        f. Motion characterization video analysis
+        g. Wind BCS video trajectory analysis
+        h. Laser-based cross-check
+        i. Earth mover's metric for comparisons (scipy has a method, also maaaaaybe look at https://www.cs.cmu.edu/~efros/courses/LBMV07/Papers/rubner-jcviu-00.pdf)
```
```suggestion
        i. Earth mover's metric for comparisons (scipy has a method, also look at https://www.cs.cmu.edu/~efros/courses/LBMV07/Papers/rubner-jcviu-00.pdf)
```

bbean23 commented on opencsp/common/lib/cv/SpotAnalysis.py:
```diff
@@ -34,85 +35,117 @@ class SpotAnalysis(Iterator[tuple[SpotAnalysisOperable]]):
+        f. Motion characterization video analysis
+        g. Wind BCS video trajectory analysis
+        h. Laser-based cross-check
+        i. Earth mover's metric for comparisons (scipy has a method, also maaaaaybe look at https://www.cs.cmu.edu/~efros/courses/LBMV07/Papers/rubner-jcviu-00.pdf)
```
Thanks for your fix on this @e10harvey !

bbean23 commented:
@e10harvey Please review again. I made some small changes in order to get the build to pass and documentation to render correctly:

* fixed docstrings in CacheableImage to be compatible with ReStructuredText
* changed "AbstractSpotAnalysisImagesProcessor" to "AbstractSpotAnalysisImageProcessor"
